firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -0,0 +1,358 @@
+    if (audioHelper != null) {
+      while (!playBackQueue.isEmpty()) playBackQueue.poll()
+      while (!audioQueue.isEmpty()) audioQueue.poll()
+      audioHelper!!.release()
+      audioHelper = null
+    }
+  }
+
+  /** Stop receiving from the server. */
+  public fun stopReceiving() {
+    if (!startedReceiving) {
+      return
+    }
+    receiveChannel.cancel()
+    receiveChannel = Channel()
+    startedReceiving = false
+  }
+
+  /**
+   * Receives responses from the server for both streaming and standard requests.
+   *
+   * @param outputModalities The list of output formats to receive from the server.
+   *
+   * @return A [Flow] which will emit [LiveContentResponse] as and when it receives it
+   *
+   * @throws [SessionAlreadyReceivingException] when the session is already receiving.
+   */
+  public suspend fun receive(outputModalities: List<ContentModality>): Flow<LiveContentResponse> {
+    if (startedReceiving) {
+      throw SessionAlreadyReceivingException()
+    }
+
+    val flowReceive = session!!.incoming.receiveAsFlow()
+    CoroutineScope(Dispatchers.IO).launch { flowReceive.collect { receiveChannel.send(it) } }
+    return flow {
+      startedReceiving = true
+      while (true) {
+        val message = receiveChannel.receive()
+        val receivedBytes = (message as Frame.Binary).readBytes()
+        val receivedJson = receivedBytes.toString(Charsets.UTF_8)
+        if (receivedJson.contains("interrupted")) {
+          emit(LiveContentResponse(null, LiveContentResponse.Status.INTERRUPTED, null))
+          continue
+        }
+        if (receivedJson.contains("turnComplete")) {
+          emit(LiveContentResponse(null, LiveContentResponse.Status.TURN_COMPLETE, null))
+          continue
+        }
+        try {
+          val functionContent = Json.decodeFromString<ToolCallSetup.Internal>(receivedJson)
+          emit(
+            LiveContentResponse(
+              null,
+              LiveContentResponse.Status.NORMAL,
+              functionContent.toolCall.functionCalls.map {
+                FunctionCallPart(it.name, it.args.orEmpty().mapValues { x -> x.value ?: JsonNull })
+              }
+            )
+          )
+          continue
+        } catch (_: Exception) {}
+        try {
+
+          val serverContent = Json.decodeFromString<ServerContentSetup.Internal>(receivedJson)
+          val data = serverContent.serverContent.modelTurn.toPublic()
+          if (outputModalities.contains(ContentModality.AUDIO)) {
+            if (data.parts[0].asInlineDataPartOrNull()?.mimeType?.equals("audio/pcm") == true) {
+              emit(LiveContentResponse(data, LiveContentResponse.Status.NORMAL, null))
+            }
+          }
+          if (outputModalities.contains(ContentModality.TEXT)) {
+            if (data.parts[0] is TextPart) {
+              emit(LiveContentResponse(data, LiveContentResponse.Status.NORMAL, null))
+            }
+          }
+        } catch (_: Exception) {}
```

Comment:
Are these supposed to be swallowing all exceptions?

Reply:
yeah supposed to be.. We have to try to decode to different json formats and see the one it decodes to.

Reply:
Hmm, is there a way to tell what the format is without doing the try/catch decoding? This seems like a very processing intensive way to do this.

Reply:
hmm we could do string matching like the way i am doing for turnComplete and pass it to different decoders. There are just two try/catch decoders though one for FunctionCalling and one for the mainResponse